### ISSUE TITLE
How to do cleanup of resources?

### ISSUE THREAD

**Issue by sergiously** [jasnell/piscina#135]
According to the documentation and to some other closed issues users have posted here, we do know there is no way to do a graceful shutdown of workers and free up resources as workers are abruptly terminated.

I'll explain my case: I've been thinking in using the **puppeteer** library to convert some .html files to .pdf format. My goal is convert hundreds of files as fast as I can, and Piscina with its workers could be perfect for it! However I'm scratching my head with one issue:

The puppeteer library spawns a Chromium instance to make the conversion. Spawning these instances takes CPU time and makes the conversion slower, so the ideal case would be **spawning a Chromium instance for each worker at initialization**. However when said workers are terminated/shutdown for being idle, the chromium instance would still remaing in memory taking up resources. **And closing these instances would be out of question, since we have no way to handle worker termination event**.

How can these instances be cleaned up on worker termination? Maybe I should switch my focus and manually close them outside of the worker?

**Comment by jasnell** [comment]
One approach for this would be to establish a separate "watch dog" thread that can terminate any orphaned chromium instances when there's no activity. This would be separate from Piscina's worker pool. When a worker spins up, it would notify the watch dog thread about the chromium instance and would send heartbeat pings to the watchdog as it runs. As long as the heartbeat pings keep coming in, the worker is active and healthy. If the watch dog stops getting the heartbeat pings, it terminates the associated chromium instance.

**Comment by sergiously** [comment]
Sorry for my late response, but that did actually work! We've managed to implement a watch dog that terminates ophaned chromium instances, and now our mass html to pdf conversion works flawlessly. Thank you very much jasnell!

Closing this issue as my question was more related to a specific implementation rather than Piscina itself. According to the documentation and other closed issues, **Piscina doesn't have a "worker termination" event, hence it's impossible to do a graceful cleanup of resources** (except for fs handlers)
